File: apps/sdk-typescript/src/crypto/codec.ts
/**
 * Encoding helpers — pure, no crypto. Inlined from `@colber/core-crypto` so
 * the SDK has no runtime workspace dependency.
 *
 * The base58btc encoder is the one required by the W3C `did:key` spec
 * (multibase `z` prefix). RFC 4648 base64 (with padding) is used for
 * signature/key wire formats. Constant-time compare is here for parity with
 * core-crypto even though the SDK doesn't currently use it.
 */

const BASE58_ALPHABET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

/** RFC 4648 base64 (with padding) — encode. */
export const toBase64 = (bytes: Uint8Array): string => Buffer.from(bytes).toString('base64');

/** RFC 4648 base64 (with padding) — decode. */
export const fromBase64 = (str: string): Uint8Array => Uint8Array.from(Buffer.from(str, 'base64'));

/**
 * Base58btc encoding (Bitcoin alphabet). Used by the multibase `z` prefix
 * required by the W3C did:key spec.
 */
export const toBase58btc = (bytes: Uint8Array): string => {
  if (bytes.length === 0) {
    return '';
  }

  // Count leading zero bytes — they map 1:1 to leading '1' chars in base58.
  let zeros = 0;
  while (zeros < bytes.length && bytes[zeros] === 0) {
    zeros++;
  }

  const size = Math.ceil(((bytes.length - zeros) * 138) / 100) + 1;
  const buffer = new Uint8Array(size);
  let length = 0;

  for (let i = zeros; i < bytes.length; i++) {
    let carry = bytes[i] ?? 0;
    let j = 0;
    for (let k = size - 1; (carry !== 0 || j < length) && k >= 0; k--, j++) {
      carry += 256 * (buffer[k] ?? 0);
      buffer[k] = carry % 58;
      carry = Math.floor(carry / 58);
    }
    length = j;
  }

  let it = size - length;
  while (it < size && buffer[it] === 0) {
    it++;
  }

  let out = '1'.repeat(zeros);
  for (; it < size; it++) {
    const idx = buffer[it];
    if (idx === undefined) {
      continue;
    }
    out += BASE58_ALPHABET[idx];
  }
  return out;
};

/** Inverse of {@link toBase58btc}. Throws on an invalid character. */
export const fromBase58btc = (str: string): Uint8Array => {
  if (str.length === 0) {
    return new Uint8Array();
  }

  let zeros = 0;
  while (zeros < str.length && str[zeros] === '1') {
    zeros++;
  }

  const size = Math.ceil(((str.length - zeros) * 733) / 1000) + 1;
  const buffer = new Uint8Array(size);
  let length = 0;

  for (let i = zeros; i < str.length; i++) {
    const ch = str[i];
    if (ch === undefined) {
      continue;
    }
    const carryStart = BASE58_ALPHABET.indexOf(ch);
    if (carryStart === -1) {
      throw new Error(`Invalid base58 character: ${ch}`);
    }

    let carry = carryStart;
    let j = 0;
    for (let k = size - 1; (carry !== 0 || j < length) && k >= 0; k--, j++) {
      carry += 58 * (buffer[k] ?? 0);
      buffer[k] = carry & 0xff;
      carry >>= 8;
    }
    length = j;
  }

  let it = size - length;
  while (it < size && buffer[it] === 0) {
    it++;
  }

  const out = new Uint8Array(zeros + (size - it));
  out.set(new Uint8Array(zeros).fill(0), 0);
  out.set(buffer.subarray(it), zeros);
  return out;
};

File: apps/sdk-python/src/colber_sdk/crypto/_codec.py
"""Encoding helpers — pure, no crypto.

Mirrors ``apps/sdk-typescript/src/crypto/codec.ts`` so the Python SDK
produces byte-identical outputs to the TS SDK on the same inputs.

- RFC 4648 base64 (with padding) for keys/signatures on the wire.
- Base58btc (Bitcoin alphabet) for the multibase ``z`` prefix required
  by the W3C did:key spec.
"""

from __future__ import annotations

import base64

BASE58_ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"


def to_base64(data: bytes) -> str:
    """RFC 4648 base64 (with padding) — encode."""
    return base64.b64encode(data).decode("ascii")


def from_base64(s: str) -> bytes:
    """RFC 4648 base64 (with padding) — decode.

    Tolerant of missing padding (some senders strip it). Raises
    ``binascii.Error`` on illegal characters.
    """
    # Auto-pad to a multiple of 4. Mirrors how Node's Buffer.from('...', 'base64')
    # tolerates unpadded inputs the TS test suite occasionally produces.
    padding = (-len(s)) % 4
    return base64.b64decode(s + ("=" * padding), validate=False)


def to_base58btc(data: bytes) -> str:
    """Base58btc encoding (Bitcoin alphabet).

    Used by the multibase ``z`` prefix required by the W3C did:key spec.
    Algorithm matches ``toBase58btc`` in the TS SDK byte-for-byte.
    """
    if len(data) == 0:
        return ""

    # Count leading zero bytes — they map 1:1 to leading '1' chars.
    zeros = 0
    while zeros < len(data) and data[zeros] == 0:
        zeros += 1

    size = ((len(data) - zeros) * 138) // 100 + 1
    buffer = bytearray(size)
    length = 0

    for i in range(zeros, len(data)):
        carry = data[i]
        j = 0
        k = size - 1
        while (carry != 0 or j < length) and k >= 0:
            carry += 256 * buffer[k]
            buffer[k] = carry % 58
            carry //= 58
            k -= 1
            j += 1
        length = j

    it = size - length
    while it < size and buffer[it] == 0:
        it += 1

    out = "1" * zeros
    while it < size:
        out += BASE58_ALPHABET[buffer[it]]
        it += 1
    return out


def from_base58btc(s: str) -> bytes:
    """Inverse of :func:`to_base58btc`. Raises ``ValueError`` on invalid char."""
    if len(s) == 0:
        return b""

    zeros = 0
    while zeros < len(s) and s[zeros] == "1":
        zeros += 1

    size = ((len(s) - zeros) * 733) // 1000 + 1
    buffer = bytearray(size)
    length = 0

    for i in range(zeros, len(s)):
        ch = s[i]
        carry_start = BASE58_ALPHABET.find(ch)
        if carry_start == -1:
            raise ValueError(f"Invalid base58 character: {ch}")

        carry = carry_start
        j = 0
        k = size - 1
        while (carry != 0 or j < length) and k >= 0:
            carry += 58 * buffer[k]
            buffer[k] = carry & 0xFF
            carry >>= 8
            k -= 1
            j += 1
        length = j

    it = size - length
    while it < size and buffer[it] == 0:
        it += 1

    out = bytearray(zeros) + buffer[it:]
    return bytes(out)
